src/rust/engine/fs/store/src/lib.rs

```diff
@@ -741,16 +748,20 @@ impl Store {
             match store.local.entry_type(digest.0).await {
               Ok(Some(EntryType::File)) => Ok(Either::Left(*digest)),
               Ok(Some(EntryType::Directory)) => {
-                // Locally expand the directory.
-                let reachable = store
-                  .into_local_only()
+                let store_for_expanding = match missing_behavior {
+                  LocalMissingBehavior::Fetch => store,
+                  LocalMissingBehavior::Error | LocalMissingBehavior::Ignore => {
+                    store.into_local_only()
+                  }
+                };
+                let reachable = store_for_expanding
                   .expand_directory(*digest)
                   .compat()
                   .await?;
                 Ok(Either::Right(reachable))
               }
               Ok(None) => {
-                if allow_missing {
+                if let LocalMissingBehavior::Ignore = missing_behavior {
```

Comment:
Style-nit: Maybe switch this to a `match` expression instead of using `let` syntax? It would improve readability since the `let` binding is not binding any names. And fits nicely with your earlier use of a `match missing_behavior`.